extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java

```diff
@@ -299,36 +301,34 @@ private static ReactiveMongoCollection mongoCollection(Object entity) {
     }
 
     private static ReactiveMongoDatabase mongoDatabase(MongoEntity entity) {
-        ReactiveMongoClient mongoClient = mongoClient(entity);
+        ReactiveMongoClient mongoClient = clientFromArc(entity, ReactiveMongoClient.class);
         if (entity != null && !entity.database().isEmpty()) {
             return mongoClient.getDatabase(entity.database());
         }
-        String databaseName = getDefaultDatabaseName();
+        String databaseName = getDefaultDatabaseName(entity);
         return mongoClient.getDatabase(databaseName);
     }
 
-    private static String getDefaultDatabaseName() {
-        if (defaultDatabaseName == null) {
-            synchronized (MongoOperations.class) {
-                if (defaultDatabaseName == null) {
-                    defaultDatabaseName = ConfigProvider.getConfig()
-                            .getValue(MONGODB_DATABASE, String.class);
+    private static String getDefaultDatabaseName(MongoEntity entity) {
+        final String beanName = beanName(entity);
+        return defaultDatabaseName.computeIfAbsent(beanName, new Function<String, String>() {
+            @Override
+            public String apply(String s) {
+                MongoClients mongoClients = Arc.container().instance(MongoClients.class).get();
+                MongoClientConfig matchingMongoClientConfig = mongoClients.getMatchingMongoClientConfig(beanName);
```

Comment:
This should use the `s` parameter (better rename it key by the way) to avoid a synthetic accessor

Reply:
Good catch